language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE ConstraintKinds   #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes        #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TypeApplications  #-}
+
+{- Note [Budgeting]
+
+Each operation costs a certain amount of memory and CPU.
+For memory Plutus counts this usage via ExMemory units, which correspond to machine words (64bit).
+For CPU, it's ExCPU, which does not have a base as just yet.
+
+First, the memory cost of the initial AST is added to the budget. See Note [Memory Usage for Plutus]. This operation currently does not cost any CPU.
+Then each machine requires a certain amount of memory and CPU.
+The builtin operations may require different amounts of budget, depending on the input size.
+
+If a computation runs out of Memory or CPU, it is aborted.
```

Comment:
Perhaps worth noting that in this case the user gets `Failure`, i.e. the same thing as if `Error` was called.